enterprise/server/remote_execution/containers/podman/podman.go

```diff
@@ -443,13 +457,87 @@ func (c *podmanCommandContainer) PullImage(ctx context.Context, creds container.
 
 	ps.mu.Lock()
 	defer ps.mu.Unlock()
+	if *imageStreamingEnabled {
+		if err := c.prepareToStreamImage(ctx); err != nil {
+			return err
+		}
+	}
 	if err := c.pullImage(ctx, creds); err != nil {
 		return err
 	}
 	ps.pulled = true
 	return nil
 }
 
+func (c *podmanCommandContainer) prepareToStreamImage(ctx context.Context) error {
+	ctx, err := prefix.AttachUserPrefixToContext(ctx, c.env)
+	if err != nil {
+		return err
+	}
+	resp, err := c.env.GetSociArtifactStoreClient().GetArtifacts(ctx, &socipb.GetArtifactsRequest{Image: c.image})
+	if err != nil {
+		return err
+	}
+	blobReq := repb.BatchReadBlobsRequest{
+		InstanceName:   "soci",
+		DigestFunction: repb.DigestFunction_SHA256,
+	}
+	for _, artifact := range resp.Artifacts {
+		if artifact.Type == socipb.Type_UNKNOWN_TYPE {
+			return status.InternalErrorf("SociArtifactStore returned unknown artifact with hash %s" + artifact.Digest.Hash)
+		} else if artifact.Type == socipb.Type_SOCI_INDEX {
+			// Write the index file, this is what the snapshotter uses to
+			// associate the requested image with its soci index.
+			if err = os.MkdirAll(sociIndexDirectory, 0644); err != nil {
+				return err
+			}
+			sociIndexDigest := godigest.NewDigestFromEncoded(godigest.SHA256, artifact.Digest.Hash)
+			imageDigest := godigest.NewDigestFromEncoded(godigest.SHA256, resp.ImageId)
+			if err = os.WriteFile(sociIndex(imageDigest), []byte(sociIndexDigest.String()), 0644); err != nil {
+				return err
+			}
+		}
+		blobReq.Digests = append(blobReq.Digests, artifact.Digest)
+	}
+	blobResp, err := c.env.GetContentAddressableStorageClient().BatchReadBlobs(ctx, &blobReq)
+	if err != nil {
+		return err
+	}
+	for _, blob := range blobResp.Responses {
+		err := gstatus.ErrorProto(blob.Status)
+		if err != nil {
+			return err
+		}
+		// Write the artifact to the local filesystem so the snapshotter can
+		// read it.
+		if err = os.MkdirAll(sociBlobDirectory, 0644); err != nil {
+			return err
+		}
+		if err = os.WriteFile(sociBlob(blob.Digest), blob.Data, 0644); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func sociIndex(digest godigest.Digest) string {
+	return sociIndexDirectory + digest.Encoded()
+}
+
+func sociBlob(digest *repb.Digest) string {
+	return sociBlobDirectory + digest.Hash
+}
+
+func resourceName(digest *repb.Digest) *rspb.ResourceName {
+	return &rspb.ResourceName{
+		Digest:         digest,
+		InstanceName:   "soci",
```

Comment:
you can remove the instance name. we used to separate CAS artifacts by instance name, but no longer do.

Reply:
Done